### PR TITLE
[12.0][MIG] account consolidation

### PR DESCRIPTION
**Pull request by i-vyshnevska**
migration and improvements of account consolidation

### CODE REVIEW

**Comment by tbaden** [comment]
runbot error:
You try to install module 'account_consolidation' that depends on module 'currency_monthly_rate'.\nBut the latter module is not available in your system.", '')
Message not found: 'Modules loaded.

File: account_consolidation/__manifest__.py
Comment (Tonow-c2c) [inline account_consolidation/__manifest__.py]:
```suggestion
# Copyright 2011-2019 Camptocamp SA
```

File: account_consolidation/views/company_consolidation_profile.xml
Comment (Tonow-c2c) [inline account_consolidation/views/company_consolidation_profile.xml]: Why add one hello ?

**Comment by i-vyshnevska** [comment]
@grindtildeath can you pass the review on this

**Comment by i-vyshnevska** [comment]
@fclementic2c functional review welcome as well

File: account_consolidation/tests/test_account_consolidation.py
Comment (sebalix) [inline account_consolidation/tests/test_account_consolidation.py]: Use `fields.Date.to_string(reversed_move.date)`

File: account_consolidation/wizard/consolidation_check_view.xml
Comment (sebalix) [inline account_consolidation/wizard/consolidation_check_view.xml]: We should keep the encoding for XML files.

Comment (i-vyshnevska) [reply account_consolidation/wizard/consolidation_check_view.xml]: in v12 no need to keep them anymore, you even will have lint complaining on it

File: account_consolidation/models/account.py
Comment (grindtildeath) [inline account_consolidation/models/account.py]: :warning: We can't force the consolidation on the same type of account.
It won't work if all the accounts are mapped to the same consolidation account, but we must be able to map between different account types. (doing it right is the user's responsibility)

File: account_consolidation/wizard/consolidation_check_view.xml
Comment (grindtildeath) [reply account_consolidation/wizard/consolidation_check_view.xml]: Can you point to an example of lint complaining ? In odoo, most of the XML files still have the encoding set...

File: account_consolidation/readme/CONFIGURE.rst
Comment (grindtildeath) [inline account_consolidation/readme/CONFIGURE.rst]: Please remove this line

File: account_consolidation/readme/CONFIGURE.rst
Comment (grindtildeath) [inline account_consolidation/readme/CONFIGURE.rst]: Is this still the case in v12.0 ?!

File: account_consolidation/models/account_move.py
Comment (grindtildeath) [inline account_consolidation/models/account_move.py]: This field definition looks like a copy-paste from https://github.com/OCA/account-financial-tools/blob/11.0/account_reversal/models/account_move.py#L15
If it's now only used to find consolidation entries to be reversed, I'd remove string and help, and add a comment like "Technical field to search consolidation entries that must be reverted on the next run (previously defined in account_reversal)" and set it readonly.

**Review by grindtildeath — CHANGES_REQUESTED** [review]
Good job :+1: just a few things to change.

File: account_consolidation/models/account.py
Comment (i-vyshnevska) [reply account_consolidation/models/account.py]: this was made to avoid issue which we have in consolidation,without domain user will that any type of account is accessible and it' not,  or I can add a help message, at least something

Comment (grindtildeath) [reply account_consolidation/models/account.py]: IMO those who want to use this module must know what consolidation is about. 
So if the balance of all the accounts mapped to the same account is 0, it's normal that no consolidation entries are generated.
Maybe you can add a note in the known issues that states "Functional mapping of accounts is not tested. If no consolidation entries are generated, it's probably because the balance of mapped accounts is 0."

File: account_consolidation/wizard/consolidation_check_view.xml
Comment (i-vyshnevska) [reply account_consolidation/wizard/consolidation_check_view.xml]: I will turn it back, can't find, but definitely  saw somewhere :(

File: account_consolidation/readme/CONFIGURE.rst
Comment (i-vyshnevska) [reply account_consolidation/readme/CONFIGURE.rst]: not very competent at this point, there is a specific test for this right, and specific  hook to remove access group (yml in previous version)
I just kept it without significant changes

Comment (grindtildeath) [reply account_consolidation/readme/CONFIGURE.rst]: The issue was that `group_multi_company` is defined in `res.config.settings`, and saving a change in Accounting > Configuration > Settings (e.g. updating consolidation profile) would remove the group for the users who had it defined, if the group is not implied by group employees.

I tried to reproduce on runbot using consolidation manager (I tried to set a rate of 50% on one of the subsidiaries) but I get the following error:

" The operation cannot be completed, probably due to the following:
- deletion: you may be trying to delete a record while other records still reference it
- creation/update: a mandatory field is not correctly set

[object with reference: company_id - company.id] "

So it seems some rights must be fixed to allow this operation.

Comment (i-vyshnevska) [reply account_consolidation/readme/CONFIGURE.rst]: @grindtildeath multicompany settings was moved to general settings so they are not conflicting any more, 
that error caused by not appropriate wizard configuration behavior, I made some improvements of it, and put them in separate commit so we can back-port it to older versions if need

File: account_consolidation/security/res_groups.xml
Comment (grindtildeath) [inline account_consolidation/security/res_groups.xml]: Can you please add `base.group_partner_manager` here ?

File: account_consolidation/models/account_move.py
Comment (grindtildeath) [inline account_consolidation/models/account_move.py]: There's a UX issue with the existing field in Odoo. Consolidation entries appear like that:
image

It's a bit confusing that odoo's checkbox appears False while this field is True but not displayed.

Suggestion:
Add the `to_be_reversed` field in the view (below reverse automatically), and add the following slots here:
```
string='To be reversed on next consolidation'
help='Entries generated by the consolidation wizard are automatically reversed on the next run of the consolidation wizard'
```

And make sure the field is only displayed to `group_consolidation_manager`

File: account_consolidation/wizard/consolidation_consolidate.py
Comment (grindtildeath) [inline account_consolidation/wizard/consolidation_consolidate.py]: There's an issue here with the moves to reverse (I'm not sure if it's only on v12.0 so maybe it's worth it to write a test and backport it to 11.0 to check if we have a bug)

Here's what I did:
1. Run a consolidation on August (ok)
2. Post the consolidation moves from August (ok)
3. Run a consolidation on October, August moves are reversed on the first day of October EDIT: ~(ok)~ (ko august moves should not be posted)
4. Run a consolidation on December without posting the moves from October. (ko)
image

On point 4, all the moves from the previous consolidation are reversed, however:
- Moves from the August consolidation are reversed a second time (they should not as they were already reversed with the consolidation from October)
- Moves from the October consolidation are reversed and posted although ~they were not posted~ (it's correct to reverse but they should stay unposted) EDIT: reversals should never be posted automatically by running the wizard

**Review by grindtildeath — CHANGES_REQUESTED** [review]
Looks quite good overall :+1: 
Two small things and a bigger issue with reversal to address and we should be good.

File: account_consolidation/models/account_move.py
Comment (i-vyshnevska) [reply account_consolidation/models/account_move.py]: I've managed to moved to the default auto_reverse flag and rid of this field, functional test seems ok, I am adding ore to test entries

**Review by grindtildeath — COMMENTED** [review]
@i-vyshnevska I'm not really sure what to think about your last change. :confused: 

Reversals are supposed to happen at the next run of the wizard with a fixed date on the first day of the month and not be planned ages before at the time the entries are generated (that's why I said "but we don't want these moves to be reversed at a specific date, we want these moves to be reversed when the consolidation wizard is run, i.e. when we generate new consolidation entries" and told you IDK if we can use Odoo's field).

Now instead of having to check only this last change, and be finished with this task, we need a functional review to know if this is an acceptable solution, before I can pass by again... :roll_eyes:

**Comment by i-vyshnevska** [comment]
@grindtildeath  I've dropped that option with dates, but I will leave tests as is, because we must be sure  that reversed moves will not be duplicated when consolidation wizard running

**Comment by grindtildeath** [comment]
@i-vyshnevska I edited https://github.com/OCA/account-consolidation/pull/19#discussion_r286548199 after the work we did on #20 .
I'll try to open a PR on your branch to fix the small issues I see.

**Comment by renatonlima** [comment]
Please check and fix pyflake8:

`======== Testing test_flake8 ========
account_consolidation/tests/test_account_consolidation.py:428:13: E122 continuation line missing indentation or outdented
account_consolidation/tests/test_account_consolidation.py:429:13: E122 continuation line missing indentation or outdented
account_consolidation/tests/test_account_consolidation.py:430:9: E122 continuation line missing indentation or outdented`

**Comment by grindtildeath** [comment]
@sebalix @Tonow-c2c Can you guys please update your reviews ? :pray:

**Review by sebalix — APPROVED** [review]

**Review by Tonow-c2c — APPROVED** [review]
LGTM Just code review

**Review by TDu — APPROVED** [review]
:+1:

**Review by grindtildeath — APPROVED** [review]

**Comment by Martronic-SA** [comment]
What is the status here? seems mergeable since june?

**Comment by fclementic2c** [comment]
Sorry but we will no longer invest time on this module since odoo did a
similar module for the enterprise version
Just to let you know...


Le mar. 12 nov. 2019 à 14:42, Martronic SA <notifications@github.com> a
écrit :

**Comment by Martronic-SA** [comment]
Not everyone runs Odoo Enterprise and now that the job has been done, why not merge it?
We will certainly find other maintainers for that module in Community Edition no?

**Comment by Martronic-SA** [comment]
@fclementic2c thank you!

**Pull request merged**
